src/DefaultBuilder/src/Configuration.cs

```diff
@@ -216,5 +285,109 @@ IEnumerator IEnumerable.GetEnumerator()
+            private readonly Configuration _config;
+            private object? _fileProvider;
+
+            public ConfigurationBuilderProperties(Configuration config)
+            {
+                _config = config;
+            }
+
+            public object this[string key]
+            {
+                get => _properties[key];
+                set
+                {
+                    _properties[key] = value;
+                    CheckForFileProviderChange(key, value);
+                }
+            }
+
+            public ICollection<string> Keys => _properties.Keys;
+
+            public ICollection<object> Values => _properties.Values;
+
+            public int Count => _properties.Count;
+
+            public bool IsReadOnly => false;
+
+            public void Add(string key, object value)
+            {
+                _properties.Add(key, value);
+                CheckForFileProviderChange(key, value);
+            }
+
+            public void Add(KeyValuePair<string, object> item)
+            {
+                ((IDictionary<string, object>)_properties).Add(item);
+                CheckForFileProviderChange(item.Key, item.Value);
+            }
+
+            public void Clear()
+            {
+                _properties.Clear();
+                CheckForFileProviderChange(FileProviderKey, null);
+            }
+
+            public bool Contains(KeyValuePair<string, object> item)
+            {
+                return _properties.Contains(item);
+            }
+
+            public bool ContainsKey(string key)
+            {
+                return _properties.ContainsKey(key);
+            }
+
+            public void CopyTo(KeyValuePair<string, object>[] array, int arrayIndex)
+            {
+                ((IDictionary<string, object>)_properties).CopyTo(array, arrayIndex);
+            }
+
+            public IEnumerator<KeyValuePair<string, object>> GetEnumerator()
+            {
+                return _properties.GetEnumerator();
+            }
+
+            public bool Remove(string key)
+            {
+                var wasRemoved = _properties.Remove(key);
+                CheckForFileProviderChange(key, null);
+                return wasRemoved;
+            }
+
+            public bool Remove(KeyValuePair<string, object> item)
+            {
+                var wasRemoved = ((IDictionary<string, object>)_properties).Remove(item);
+                CheckForFileProviderChange(item.Key, null);
+                return wasRemoved;
+            }
+
+            public bool TryGetValue(string key, [MaybeNullWhen(false)] out object value)
+            {
+                return _properties.TryGetValue(key, out value);
+            }
+
+            IEnumerator IEnumerable.GetEnumerator()
+            {
+                return _properties.GetEnumerator();
+            }
+
+            private void CheckForFileProviderChange(string key, object? value)
```

Comment:
Surprised we special-cased this to the file provider. I thought we would invalidate the configuration on every property change because it's impossible to know how they would affect the resulting configuration.

Reply:
This isn't even checking for a change to the configuration per se, but `IConfigurationBuilder.Properties`. This is very likely to change where file-based config provider are loaded from.

As for config effecting where other sources are loaded from, that's certainly possible in a scenario like the one laid out as a "Usage Example" in https://github.com/dotnet/runtime/issues/51770, but I think it works fine without reloading sources as long as you're careful about the order you load sources in.

If in the future we really want this to be able to define config-dependent config sources that react to *new* config, I think we'd have to add an API like. `IDisposable Configuration.RegisterDependentSource(Action<IConfigurationBuilder> dependentConfigCallback)`.

Ex:

``` C#
    using var config = new Configuration();

    config.AddEnvironmentVariables(prefix: "MyCustomPrefix_");

    // If MyCustomPrefix_FileConfig = "enabled", this will add config from MyConfig.json
	cofig.RegisterDependentSource(emptyConfigBuilder =>
    {
        if (config["FileConfig"] == "enabled")
        {
            emptyConfigBuilder.AddJsonFile("MyConfig.json", optional: true, reloadOnChange: true);
        }
    });
    
    // Adding more sources or any change token firing causes the RegisterDependentSource to be rerun.
    // If MyCustomPrefix2_FileConfig = "disabled", this will remove config from MyConfig.json even if MyCustomPrefix_FileConfig = "enabled"
    config.AddEnvironmentVariables(prefix: "MyCustomPrefix2_");

    string myValueFromJson = config["JsonConfigValue"];
    // ...
```


Though that'd still be tricky, because how do you know it's stable? How do you know `dependentConfigCallback` didn't add a source that would cause it to add a different source the next time? If there are multiple callbacks, do you just run them in order?

Reply:
Nvm. I reread the comment and see you are referring specifically to `IConfigurationBuilder.Properties` @pakrym. This makes way more sense. I'll update to reload everything on any property change.